sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FailureSafeParser.scala

```diff
@@ -56,9 +57,14 @@ class FailureSafeParser[IN](
     }
   }
 
+  private val skipParsing = optimizeEmptySchema && schema.isEmpty
   def parse(input: IN): Iterator[InternalRow] = {
     try {
-      rawParser.apply(input).toIterator.map(row => toResultRow(Some(row), () => null))
+     if (skipParsing) {
+       Iterator.single(InternalRow.empty)
```

Comment:
nit: `Iterator.empty`

Reply:
It is not the same. If you return empty iterator, `count()` will always return `0`.

Reply:
ohh yes my bad!